data_structures/src/chain.rs

```diff
@@ -784,6 +882,236 @@ pub enum InventoryItem {
+                }
+            }
+            DataRequestStage::TALLY => {
+                if self.info.tally.is_none() {
+                    DataRequestStage::TALLY
+                } else {
+                    panic!("Data request in tally stage should have been removed from the pool");
+                }
+            }
+        };
+
+        self.stage != old_stage
+    }
+}
+
+/// Data request current stage
+#[derive(Copy, Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
+pub enum DataRequestStage {
+    /// Expecting commitments for data request
+    COMMIT,
+    /// Expecting reveals to previously published commitments
+    REVEAL,
+    /// Expecting tally to be included in block
+    TALLY,
+}
+
+/// Pool of active data requests
+#[derive(Clone, Debug, Default, PartialEq, Serialize, Deserialize)]
+pub struct ActiveDataRequestPool {
+    /// Current active data request, in which this node has announced commitments.
+    /// Key: Data Request Pointer, Value: Reveal Transaction
+    pub waiting_for_reveal: HashMap<OutputPointer, Transaction>,
+    /// List of active data request output pointers ordered by epoch (for mining purposes)
+    pub data_requests_by_epoch: BTreeMap<Epoch, HashSet<OutputPointer>>,
+    /// List of active data requests indexed by output pointer
+    pub data_request_pool: HashMap<OutputPointer, DataRequestState>,
+    /// List of data requests that should be persisted into storage
+    pub to_be_stored: Vec<(OutputPointer, DataRequestReport)>,
+    /// Cache which maps commit_pointer to data_request_pointer
+    /// and reveal_pointer to data_request_pointer
+    pub dr_pointer_cache: HashMap<OutputPointer, OutputPointer>,
+}
+
+pub type UnspentOutputsPool = HashMap<OutputPointer, Output>;
+
+pub type Blockchain = BTreeMap<Epoch, HashSet<Hash>>;
+
+/// Blockchain state (valid at a certain epoch)
+#[derive(Debug, Default, Serialize, Deserialize, PartialEq, Clone)]
+pub struct ChainState {
+    /// Blockchain information data structure
+    pub chain_info: Option<ChainInfo>,
+    /// Unspent Outputs Pool
+    pub unspent_outputs_pool: UnspentOutputsPool,
+    /// Data request pool
+    pub data_request_pool: ActiveDataRequestPool,
+}
+
+impl ChainState {
+    /// Method to check that all inpunts point to unspend output
+    pub fn find_unspent_outputs(&self, inputs: &[Input]) -> bool {
+        inputs.iter().all(|tx_input| {
+            let output_pointer = tx_input.get_output_pointer_from_input();
+
+            self.unspent_outputs_pool.contains_key(&output_pointer)
+        })
+    }
+    /// calculate output pointed from input
+    pub fn get_output_from_input(&self, input: &Input) -> Option<&Output> {
+        let output_pointer = input.get_output_pointer_from_input();
+
+        self.unspent_outputs_pool.get(&output_pointer)
+    }
+    /// calculate output vector from inputs vector
+    pub fn get_outputs_from_inputs(&self, inputs: &[Input]) -> Result<Vec<Output>, Input> {
```

Comment:
This can potentially be refactored as something similar to this:
```rust
let v = inputs.i.iter()
    .map(self.get_output_from_input)
    .fuse() // Make this iterator short-circuit as soon as the `.map` returns` None`
    .flatten() // Unwrap all `Option`s
    // At this point you may need `.cloned()` or `.deref()`
    .collect();
```
Feel free to mark the refactor as `TODO`.